### ISSUE TITLE
Build a library of iterator tools

### ISSUE THREAD

**Issue by ben-albrecht** [chapel-lang/chapel#12657]
It would be useful to provide a toolkit of common serial and parallel iterators to users through a library. Python’s itertools library would make a good reference for target functionality as well as early performance comparisons. However, this task is not limited to iterators available in itertools.

Assuming we're porting python [itertools](https://docs.python.org/3/library/itertools.html), we would need to determine the following for each iterator:

- Is this functionality already available in the base language?
- Is this iterator parallelizable?
  - Is it embarrassingly parallel or more complex?
- Can this iterator operate on distributed data?

There are a number of example recipes provided in the [itertools documentation](https://docs.python.org/3/library/itertools.html#itertools-recipes) that demonstrate combining multiple iterators together to create powerful constructs. Being able to reproduce many (or all) of these constructs would be a good design goal. However, note that there are some constraints that Chapel will have compared to python, due to being a statically typed language.

**Comment by ben-albrecht** [comment]
This is up for grabs as a [GSoC 2019 project](https://chapel-lang.org/gsoc/ideas.html#iterator-library). Some suggestions to GSoC applicants:

- Determining the subset of iterators to implement and which ones are parellizable/distributable is left as an exercise for the applicants, and can be included as part of the application project plan / timeline.
- Implementing one of the iterator functions, with tests, documentation, and
  performance testing in advance would be helpful in several ways:
    - It will give you an idea of how long this process takes and should give
      you a more accurate estimate for your timeline.
    - It will also show that are ramped up on interacting with Chapel's
      testing system (`start_test`) and `chpldoc`.
- I would expect that handling some parallel and/or distributed iterators would be part of the GSoC timeline. 
  - This may limit how many iterators get implemented in the summer, and that's fine. It's OK to leave some (or many) iterators out of the timeline or push them into stretch goals section.

**Comment by ben-albrecht** [comment]
Though not explicitly listed, this may be considered part of https://github.com/chapel-lang/chapel/issues/6329

**Comment by akshansh2000** [comment]
> * Determining the subset of iterators to implement and which ones are parellizable/distributable is left as an exercise for the applicants, and can be included as part of the application project plan / timeline.

I want to take this up as a GSoC 2019 project. I was just wondering if beginning with implementing the `repeat` and `product` itertools of python would be a good way to start. As far as I know, these tools are not available in Chapel as of now. Please let me know if I missed something, and whether to go ahead with this.

**Comment by ben-albrecht** [comment]
Hi @akshansh2000, `repeat` and `product` seem like good starts. I expect `repeat` to be relatively simple, but `product` will be a little more challenging.

Regarding your `product` implementation, there are a few design goals to consider:

- We'd want to support [variadic arguments](https://chapel-lang.org/docs/primers/varargs.html)
- We'd want our arguments to be generic so that we don't have to write an overload for every iterable type
- We'd want to support a repeat argument similar to [itertools.product](https://docs.python.org/3/library/itertools.html#itertools.product)
- Note that we won't be able to support heterogenous types for our arguments, due to being unable to iterate over or index into a tuple of heterogenous types in chapel. 
  - See https://github.com/chapel-lang/chapel/issues/9607 for why this is, if curious.

Beyond that, documentation, evaluating performance, and determining if the implementation should/can be parallelized would be your next steps (in any order you prefer).

**Comment by akshansh2000** [comment]
Thank you, @ben-albrecht. I'm working on it. I'll keep you updated.

**Comment by akshansh2000** [comment]
The basic implementation for the repeat tool is given:
```
iter repeat (arg, times = 0) {
    if times == 0 then
        for count in 1.. do yield arg;
    else
        for count in 1..#times do yield arg;
}

// For integer/real/complex etc..

writeln(repeat(10, 15));

// For arrays/strings

writeln(repeat([1, 4, 6, 7], 4));
writeln(repeat("Testing", 5));

// For dictionaries

var dictDomain: domain(string);
var dict: [dictDomain] int;

dict["one"] = 1;
dict["two"] = 2;
dict["three"] = 3;

writeln(repeat((dictDomain, dict), 3));

// For infinite repetition

write(repeat("infinite"));
```
Output:
```
10 10 10 10 10 10 10 10 10 10 10 10 10 10 10
1 4 6 7 1 4 6 7 1 4 6 7 1 4 6 7
Testing Testing Testing Testing Testing
({two, three, one}, 2 3 1) ({two, three, one}, 2 3 1) ({two, three, one}, 2 3 1)
infinite infinite infinite infinite .....
```

Please let me know what all I need to change in this. This sure could be parallelized, and I'll work on that next. The documentation and performace evaluation is further in the queue.

**Comment by LouisJenkinsCS** [comment]
@akshansh2000 You may want to make this into its own Pull Request. (also side note: If you add write the code block as \`\`\`chpl you get syntax highlighting)

```chapel
iter repeat (arg, times = 0) {
    if times == 0 then
        for count in 1.. do yield arg;
    else
        for count in 1..#times do yield arg;
}
```

You also will want a parallel iterator. I'll do this one for you so you get the gist of how to do so in Chapel...

```chpl
iter repeat (arg, times = 0, param tag : iterKind) where tag == iterKind.standalone {
    if times == 0 then
        forall count in 1.. do yield arg;
    else
        forall count in 1..#times do yield arg;
}
```

You'll also definitely want a leader-follower version so you can do something like this...

```chpl
forall (ix, a) in zip(repeat(fixedIdx), arr) {
   a = ix;
}
```

In the above example, you zip over an infinite stream `repeat(fixedIdx)` and some finite container.

**Comment by akshansh2000** [comment]
Thanks for the clarification of the parallel iterator, @LouisJenkinsCS! I'll work on the leader-follower part and share my progress.

(Thanks for the syntax highlighting part as well hahaha)

Also, as for the PR, I'm supposed to make a Mason package and make a PR to the mason-registry repository, right?

**Comment by LouisJenkinsCS** [comment]
Hm... I'm not sure, honestly. Previously students just submit a PR to the repository and it gets stuck somewhere (likely labeled as some kind of prototype/test), but I suppose it makes sense if @ben-albrecht would want you to implement it as a Mason package.

**Comment by akshansh2000** [comment]
Alright, I'll wait for @ben-albrecht sir to reply, and work on my code until then.

Thank you!

**Comment by ben-albrecht** [comment]
Mason package vs package module is still an open question at this point. Though, I'm leaning towards a package module right now so that we can utilize the nightly performance testing. Once mason packages have this functionality, we could migrate this to a mason package, like we plan to do for other existing package modules.

**Comment by akshansh2000** [comment]
The repeat iterator is almost complete, I believe.
Please take a look at my implementation and let me know if I need to change something.

Also, I'm a bit confused here. How do I implement infinite repetition for a parallel loop?
In the case of a `forall` loop, it needs to know to the total number of tasks to divide the `repeat` task into different tasks, while in the case of a `coforall` loop, the program terminates at runtime due to excessive memory usage. Please help me out with this, I'm a bit lost.

Here's the code; I've commented the non working code below.

```chpl
// Serial Iterator

iter repeat (arg, times = 0) {
    if times == 0 then
        for count in 1.. do yield arg;
    else
        for count in 1..#times do yield arg;
}

// Standalone Parallel Iterator

iter repeat (param tag : iterKind, arg, times = 0) where tag == iterKind.standalone {
    if times == 0 then
        coforall count in 1.. do yield arg; // parallel infinite loop creates problem
    else
        coforall count in 1..#times do yield arg;
}

// Procedure to compute chunks which are to be iterated through

proc computeChunk(r : range, myChunk, numChunks) where r.stridable == false {
    const numElems = r.length;
    const elemsPerChunk = numElems / numChunks;
    const mylow = r.low + (elemsPerChunk * myChunk);

    if (myChunk != numChunks - 1) then
        return mylow..#elemsPerChunk;
    else
        return mylow..r.high;
}

const numTasks = here.maxTaskPar; // max number of tasks supported by locale

// Leader

iter repeat (param tag : iterKind, arg, times = 0) where tag == iterKind.leader {
    if times == 0 then
        coforall task_id in 0.. do yield(0..1,); // parallel infinite loop creates problem
    else
        coforall task_id in 0..#numTasks {
            const working_iters = computeChunk(0..#times, task_id, numTasks);
            yield(working_iters,);
        }
}

// Follower

iter repeat (param tag : iterKind, arg, times = 0, followThis)
        where tag == iterKind.follower && followThis.size == 1 {
    const working_iters = followThis(1);

    for idx in working_iters do yield arg;
}

for element in repeat(1, 10) do write(element, ' '); // works - 1 1 1 1 1 1 1 1 1 1
writeln();

forall element in repeat('abc', 4) do write(element, ' '); // works - abc abc abc abc
writeln();

forall element in repeat(123) do write(element, ' '); // doesn't work, parallel infinite loop
writeln();

var arr: [1..5] int;

forall (idx, element) in zip(repeat(3, 5), arr) do element = idx; // works
writeln(arr); // - 3 3 3 3 3

forall (idx, element) in zip(repeat(2), arr) do element = idx; // doesn't work, parallel infinite loop
writeln(arr);
```

Thank you!

**Comment by ben-albrecht** [comment]
> Also, I'm a bit confused here. How do I implement infinite repetition for a parallel loop?

This is not supported today in Chapel. As a result, we will not be able to pursue parallel implementations of any infinite iterators for this project.

**Comment by akshansh2000** [comment]
Alright, so should I put the rest of the code into a module and create a PR?

**Comment by LouisJenkinsCS** [comment]
Oh yeah, I forgot that you can't `break` from a `forall` loop. I'm wondering if its wise to return a 'killswitch' which the leader and follower will check to see if it should continue.

**Comment by LouisJenkinsCS** [comment]
Note that if I make the array the 'leader', it works fine. It has to do with the leader not knowing when to stop, which is definitely a problem. [TIO](https://tio.run/##rVVRT9swEH7Pr7gJiSabCWSMFyqepk2atgek8TZNyE3cxMKxo0uyUtB@e3d26qSmwMZEq6qN/d333Xd3dvOKN0JtNpHsBAKKRvAOYo4lg07WooULOEngPgJ6yaVfo0XoKqHdsn0tDUJuet2B1JClKRQG1lKoAohq7mBCteJJ/MFAHEb9jqLjY/jecV1wZbSAS45cKaHgCyXLO4NRmHZD@zV0vIRzsBtfpS4YPDSzqgQKhyIfHpa2k8xf3eaG8qdMnrEMlHnj05V6KTUpgTKmgRwpWeJt0CyUqPeL8yj9MxW6RJOLoremDMXUTU9SedXrm5bMyrwisNtbCOeX1AsyhKYvqyiiNHIf9dEGxUjlQ65LwaBeuyUGuq/dr9bXD6liKAtOFmyFlpwcbCuXG912NuIT2bNFx1QJXXbVfGdb2L1LgY6VMCP8eNLaxddrZVYDF32/gzgkeOtTTeaRb1@8XYI3FxMnHEGWhO1E0fWoBwUqdED8yOyGcEwrWVauFaPvK97eWN@2UGnNb@0zza4bCnq0kAWNraEJc8i2bxqDtiuLNc1IzpVwjf0meCFec8iVI3zBgNv8rmVhZ/BkZ8RjeshY8jpTHmocjPW7H6HTFKwM3khdXltH1mkwtyf@lDDPycZuJPOAbXARsLEdyO/tyfpsFDX5v1rA6I6zwVeVbJOR@PG2LLcycHi4E5W28s6drSw4Vw9LMAXEmR9@e73K4taWNITvXR8WaQdeDBfN4DDOGGQ0RIReIQXGWwSDGcyGrltad5b235GLUTq22Wx7vC8x44t8xuDDP6kQdvcTKPziSG7wHH7QLXn2kyS6SXfipHIkVv5ONvT3hsznccrgLHFJ@BwvLHaSH8UoasjqCE79@wVC759SKYxo9WxoLHviOIVJbDZ/AA)

**Comment by akshansh2000** [comment]
> Note that if I make the array the 'leader', it works fine. It has to do with the leader not knowing when to stop, which is definitely a problem. [TIO](https://tio.run/##rVVRT9swEH7Pr7gJiSabCWSMFyqepk2atgek8TZNyE3cxMKxo0uyUtB@e3d26qSmwMZEq6qN/d333Xd3dvOKN0JtNpHsBAKKRvAOYo4lg07WooULOEngPgJ6yaVfo0XoKqHdsn0tDUJuet2B1JClKRQG1lKoAohq7mBCteJJ/MFAHEb9jqLjY/jecV1wZbSAS45cKaHgCyXLO4NRmHZD@zV0vIRzsBtfpS4YPDSzqgQKhyIfHpa2k8xf3eaG8qdMnrEMlHnj05V6KTUpgTKmgRwpWeJt0CyUqPeL8yj9MxW6RJOLoremDMXUTU9SedXrm5bMyrwisNtbCOeX1AsyhKYvqyiiNHIf9dEGxUjlQ65LwaBeuyUGuq/dr9bXD6liKAtOFmyFlpwcbCuXG912NuIT2bNFx1QJXXbVfGdb2L1LgY6VMCP8eNLaxddrZVYDF32/gzgkeOtTTeaRb1@8XYI3FxMnHEGWhO1E0fWoBwUqdED8yOyGcEwrWVauFaPvK97eWN@2UGnNb@0zza4bCnq0kAWNraEJc8i2bxqDtiuLNc1IzpVwjf0meCFec8iVI3zBgNv8rmVhZ/BkZ8RjeshY8jpTHmocjPW7H6HTFKwM3khdXltH1mkwtyf@lDDPycZuJPOAbXARsLEdyO/tyfpsFDX5v1rA6I6zwVeVbJOR@PG2LLcycHi4E5W28s6drSw4Vw9LMAXEmR9@e73K4taWNITvXR8WaQdeDBfN4DDOGGQ0RIReIQXGWwSDGcyGrltad5b235GLUTq22Wx7vC8x44t8xuDDP6kQdvcTKPziSG7wHH7QLXn2kyS6SXfipHIkVv5ONvT3hsznccrgLHFJ@BwvLHaSH8UoasjqCE79@wVC759SKYxo9WxoLHviOIVJbDZ/AA)

Noted. So for now should I exclude the infinite-parallel iteration from the code and make a module out of the rest of my code, and throw an error when trying to use infinite iteration with a parallel loop?

Also, should I open an issue regarding the above?

**Comment by LouisJenkinsCS** [comment]
You could halt when you have an infinite parallel loop for now, sure. The lack of a break statement in forall is a well-known issue, so no issue needed (unless Ben thinks differently)

**Comment by ben-albrecht** [comment]
Despite being well-known, I think it's useful to have an issue to reference / search for. I've created a simple one here: https://github.com/chapel-lang/chapel/issues/12700

**Comment by akshansh2000** [comment]
@ben-albrecht, @LouisJenkinsCS, I completed the serial version of the _cycle itertool_, which returns separate elements from an iterable, eg:
```chpl
writeln(cycle('ABC'));
writeln(cycle(1..4, 2));
```
```
>>  A B C A B C A B C ...
>>  1 2 3 4 1 2 3 4
```
I was working on the parallel version, but I am not sure if it makes sense to parallelize. I don't think that a shuffling of order would be desirable.

Is there some case that I'm missing where the shuffling of order might be used?

**Comment by ben-albrecht** [comment]
I don't think a standalone parallel iterator makes sense for `cycle`, but a leader/follower parallel iterator might be useful in cases where zipped order matters, e.g.

```chpl
forall (dayOfMonth, dayOfWeek) in zip(1..30, cycle('MWTRFSU')) {
  April[dayOfMonth] = dayOfWeek;
}
```

**Comment by akshansh2000** [comment]
@ben-albrecht, @LouisJenkinsCS, I am having some problem implementing the parallel version of the `cycle` tool. 

The serial version is working fine. However, the parallel version throws a `zippered iterations have non-equal lengths` error when the length of the iterable is more than 1. Here is the code for the leader-follower iterator:
```chpl
use RangeChunk;

iter cycle(param tag: iterKind, arg, times = 0) throws
    where tag == iterKind.leader {

  var numTasks = if dataParTasksPerLocale > 0 then dataParTasksPerLocale else here.maxTaskPar;

  if numTasks > times then numTasks = times;

  if __primitive("method call resolves", arg, "these") {  // to check if `arg` is iterable
    coforall tid in 0..#numTasks {
      const working_iters = chunk(0..#times, numTasks, tid);
      yield(working_iters,);
    }
  } else
    throw new owned IllegalArgumentError(
        "non-iterable type argument passed");

}

iter cycle(param tag: iterKind, arg, times = 0, followThis)
    where tag == iterKind.follower && followThis.size == 1 {

  const working_iters = followThis(1);

  for working_iters do
    for element in arg do
      yield element;

}
```
On running the following code,
```chpl
forall (el, id) in zip(cycle('ABCD', 2), 1..#8) do writeln(el, ' ', id);
```
the error for unequal lengths as described above is thrown. However, on running
```chpl
forall (el, id) in zip(cycle(['ABCD'], 8), 1..#8) do writeln(el, ' ', id);
```
the program works fine.

What could I be missing? Thanks.

**Comment by LouisJenkinsCS** [comment]
@akshansh2000 

`['ABCD']` creates an array literal (I.E if you do `writeln(arg.type : string)` you'll see this fact), while `'ABCD'` creates a string. When you do `['ABCD', 'ABCD', 'ABCD']` you get the same error; the issue has to do with the fact that you are yielding more than requested. 

`cycle('ABCD', 2)` should print out `'AB'`, while `['ABCD']` should print out `'ABCD', 'ABCD'`. To do this, you'll want to ensure that you short-circuit the computations. I.E insert a `break` inside of that follower sequential `for` loop when you hit the upper limit early.

**Comment by akshansh2000** [comment]
@LouisJenkinsCS 

Actually, `cycle('ABCD', 2)` should print `A B C D A B C D`, and that's why in the first example I set the `times` argument to 2, and let `id` iterate from `1..8`, due to the string length being 4.

One `cycle` iteration means going through the whole iterable exactly once. Sorry for the missing explanation.

**Comment by LouisJenkinsCS** [comment]
[TIO](https://tio.run/##fVLBbtswDL3nKwgPaGzMc5OdhhYp0HY7DNuhGHpPFZuxhciSIcnx0iDfnpGq7SbAGh8CRHx8JN97eSUaVMdj6xD@CF3iY9Xqze1kIj1ayHe5wljYMgUva3SwgFkKjbCiBi/KG2DUL6mLBHxlTecmQF9XoUWuw2IxIjKFoiDKPUE6S69KM3Hmdw3CDThvpS6TW6puhQXd1s/CbXigXEMhvHgSNrw8of1tcqEQ7mBGU1F/UEZFJ/EmWS3@co0gdBcw4Uh/198VeE6GhtcBvVw2VtbSyy3GUY2@MgXQCAUWnVFbdFEKQaOIaBxGCewBrq/BG8grzDfM8UKAF5Au6CFWCoNSuVkby0xeFiA1zLLs07jFPkAYpJ2HztgNKbTkft4wZ59ibgi7puP2bFURhORvJ1EV8Vlz2hcP9HsIMoW/wUDQ2IHpNBbwUykshbq3ZVuj9j@sNTbuWQEibfSX4RYIJooeSflwDouIxkwOF4O0NkqZ7rmS7v@hupCmt1Zivro6ocmcfEXGzYN6k4/Ee@@I50GNIZFnwDGNbdOgfTCtLqj3DBIGDrDleBq/KLFCBab1nEjT8JBKsla2RShMuI3cJwMwqEb2kz6j6z3ZZ7rlzMsBPzxyPvu5i9NFQ6JXFsXmfYfB94u2UHqp3OcyRpUCxYm3e5VN/NYwvX94/D5N4WuSwpwS@C2hg0YNuWUK09BHETge/wE)

```chpl
use RangeChunk;

iter cycle(arg, times = 0, param tag: iterKind) throws
    where tag == iterKind.leader {
  writeln(arg.type : string);
  var numTasks = if dataParTasksPerLocale > 0 then dataParTasksPerLocale else here.maxTaskPar;

  if numTasks > times then numTasks = times;

  if __primitive("method call resolves", arg, "these") {  // to check if `arg` is iterable
    coforall tid in 0..#numTasks {
      const working_iters = chunk(0..#times, numTasks, tid);
      yield(working_iters,);
    }
  } else
    throw new owned IllegalArgumentError(
        "non-iterable type argument passed");

}

iter cycle(arg, times = 0, followThis, param tag: iterKind)
    where tag == iterKind.follower && followThis.size == 1 {
  
  const working_iters = followThis(1);
  writeln(working_iters);
  var upperBound = working_iters.size;
  var _times = 0;
  label outerLoop
  while true do
    for element in arg {
      _times += 1;
      yield element;
      if _times == upperBound then break outerLoop;
    }

}

iter cycle(arg, times = 0) {}

forall (el, id) in zip(cycle('ABCD', 2), 1..#8) do writeln(el, ' ', id);
```

If you want it to go around in a cycle, it'd have to be a bit more complicated than what we have right now. If this is what you want you should obtain the `size` of the `arg` (assert that it is resolvable just like how you did for `these`), and then have some additional logic to skip the first N iterations over `arg`. RIght now it prints `'A' 'A'` because both begin at the same letter. Then again, you could make this more efficient by only creating tasks if the number of `times` exceeds `arg.size`.

**Comment by LouisJenkinsCS** [comment]
I think I misunderstood the intention here. A string by itself is iterable and yields characters; if you want the string itself to be iterated over, the array literal is the best approach.

**Comment by akshansh2000** [comment]
@LouisJenkinsCS, does that mean storing the string as an array of characters?

**Comment by LouisJenkinsCS** [comment]
[TIO](https://tio.run/##fVPBjtowEL3nK0aptCRVGqCnahEr7W57qNrDqto768ZDYuHYke2Qsohvp2OTpLDqkgMKnvfeeN68FBVrUB6PrUX4xVSJj1WrNosoEg4NFLtCYsJMmYETNVpYwiyDhhlWg2PlLXjUD6F4Cq4yurMR0NNVaNDXYbkcEblExklyT5DO0KlUXjh3uwbhFqwzQpXpgqpbZkC19TOzG99QrIEzx56YCSdPaH7qgkmEO5hRV1TvlFHSSP4mec3@@BpBaC7wgqP8XT9X0DlrGk4H9GrVGFELJ7aYxDW6SnOgFhIMWi23aOMMgkcxyViMU9gDTKfgNBQVFhuv8UKAFxA2@MF@SwxOFXqtjVdygoNQMMvzD@Mt9gHiQco66LTZkEMrzz/dkEfw0ffNrXhFIg6vi563Eyh5csHL0lPxQL@H4FD4G3YHCjvQnUIO36XEksl7U7Y1KvfNGG2SXhUgVlp9GsaAsD/WIyka1iKPqU10uJqhtZZSd8@VsP/P05UgnaikfHNzJhNG97h5MC56z7d/jGQe3BjCeAEcg9g2DZoH3SpO3AvI6LWHrcbRFqfmtNc3cOoK03FfkOdv6pUoq3MA18EDL4QSg7uUECqPwQgLHorD1vv9XrWfAkrlPnoJygwEfcGk/iqa5ESY3D88fp1k8DnNYE7h@pLShUavPGUCk8CjVR@PfwE) That should work

**Comment by akshansh2000** [comment]
That is a great approach, @LouisJenkinsCS.

I'll try to optimize it and see if we can get it done using even fewer resources. I'll get back here if I have any doubts. Thank you!

**Comment by ben-albrecht** [comment]
> TIO That should work

Nice example @LouisJenkinsCS.

I have a small suggestion - [`Reflection.canResolveMethod`](https://chapel-lang.org/docs/master/modules/standard/Reflection.html#Reflection.canResolveMethod) would be preferred over the `__primitive` call, which is not a public interface (and is therefore subject to change in the future without proper deprecation).

e.g.

```chpl
use Reflection only;

...

if Reflection.canResolveMethod(arg, "these") {
  ...
} else {
   throw new owned IllegalArgumentError(...);
}

**Comment by akshansh2000** [comment]
@LouisJenkinsCS @ben-albrecht 
> [TIO](https://tio.run/##fVPBjtowEL3nK0aptCRVGqCnahEr7W57qNrDqto768ZDYuHYke2Qsohvp2OTpLDqkgMKnvfeeN68FBVrUB6PrUX4xVSJj1WrNosoEg4NFLtCYsJMmYETNVpYwiyDhhlWg2PlLXjUD6F4Cq4yurMR0NNVaNDXYbkcEblExklyT5DO0KlUXjh3uwbhFqwzQpXpgqpbZkC19TOzG99QrIEzx56YCSdPaH7qgkmEO5hRV1TvlFHSSP4mec3@@BpBaC7wgqP8XT9X0DlrGk4H9GrVGFELJ7aYxDW6SnOgFhIMWi23aOMMgkcxyViMU9gDTKfgNBQVFhuv8UKAFxA2@MF@SwxOFXqtjVdygoNQMMvzD@Mt9gHiQco66LTZkEMrzz/dkEfw0ffNrXhFIg6vi563Eyh5csHL0lPxQL@H4FD4G3YHCjvQnUIO36XEksl7U7Y1KvfNGG2SXhUgVlp9GsaAsD/WIyka1iKPqU10uJqhtZZSd8@VsP/P05UgnaikfHNzJhNG97h5MC56z7d/jGQe3BjCeAEcg9g2DZoH3SpO3AvI6LWHrcbRFqfmtNc3cOoK03FfkOdv6pUoq3MA18EDL4QSg7uUECqPwQgLHorD1vv9XrWfAkrlPnoJygwEfcGk/iqa5ESY3D88fp1k8DnNYE7h@pLShUavPGUCk8CjVR@PfwE) That should work

This code doesn't work if the `times` argument is more than 2. [TIO1](https://tio.run/##fVPBjtowEL3nK0aptCRVGqC9LWKl3W0PVXtYVXtn3XhILBw7sh1SFvHtdGySFFZdckDB894bz5uXomINyuOxtQi/mCrxsWrVZhFFwqGBYldITJgpM3CiRgtLmGXQMMNqcKy8BY/6IRRPwVVGdzYCeroKDfo6LJcjIpfIOEnuCdIZOpXKC@du1yDcgnVGqDJdUHXLDKi2fmZ24xuKNXDm2BMz4eQJzU9dMIlwBzPqiuqdMkoayd8kr9kfXyMIzQVecJS/6@cKOmdNw@mAXq0aI2rhxBaTuEZXaQ7UQoJBq@UWbZxB8CgmGYtxCnuA6RSchqLCYuM1XgjwAsIGP9hvicGpQq@18UpOcBAKZnn@YbzFPkA8SFkHnTYbcmjl@acb8gg@@r65Fa9IxOF10fN2AiVPLnhZeioe6PcQHAp/w@5AYQe6U8jhu5RYMnlvyrZG5b4Zo03SqwLESqtPwxgQ9sd6JEXDWuQxtYkOVzO01lLq7rkS9v95uhKkE5WUb27OZMLoHjcPxkXv@faPkcyDG0MYL4BjENumQfOgW8WJewEZvfaw1Tja4tSc9voGTl1hOu4L8vxNvRJldQ7gOnjghVBicJcSQuUxGGHBQ3HYer/fq/ZTQKncRy9BmYGgL5jUX0WTnAiT@4fHr5MMvqQZzClc888p3Wg0y3MmMAlE2vXx@Bc), [TIO2](https://tio.run/##fVPBjtowEL3nK0aptCRVGkCqeljESrvbHqr2sKr2zrrxkFg4dmQ7pCzi2@nYJCmsuuSAgue9N543L0XFGpTHY2sRfjFV4mPVqs0iioRDA8WukJgwU2bgRI0WljDLoGGG1eBYeQse9UMonoKrjO5sBPR0FRr0dVguR0QukXGS3BOkM3QqlRfO3a5BuAXrjFBluqDqlhlQbf3M7MY3FGvgzLEnZsLJE5qfumAS4Q5m1BXVO2WUNJK/SV6zP75GEJoLvOAof9fPFXTOmobTAb1aNUbUwoktJnGNrtIcqIUEg1bLLdo4g@BRTDIW4xT2ANMpOA1FhcXGa7wQ4AWEDX6w3xKDU4Vea@OVnOAgFMzy/MN4i32AeJCyDjptNuTQyvNPN@QRfPR9cytekYjD66Ln7QRKnlzwsvRUPNDvITgU/obdgcIOdKeQw3cpsWTy3pRtjcp9M0abpFcFiJVWn4YxIOyP9UiKhrXIY2oTHa5maK2l1N1zJez/83QlSCcqKd/cnMmE0T1uHoyL3vPtHyOZBzeGMF4AxyC2TYPmQbeKE/cCMnrtYatxtMWpOe31DZy6wnTcF@T5m3olyuocwHXwwAuhxOAuJYTKYzDCgofisPV@v1ftp4BSuY9egjIDQV8wqb@KJjkRJvcPj18nGXxOM5hTuOZfUrrRaJbnTGASiLTr4/Ev)

I was trying to debug it.
I took another example of a simple iterator that yields `1` eight times. The serial part of it ([TIO3](https://tio.run/##S85ILEjN@f8/syS1SCExKVlDU6GaS0EhLb9IwVBPz0IhJR/Ig/CzFDLzQIKGMEEFhcrM1JwUhSxrrlourvIioBk5eRpgQzSt//8HAA)) is given as
```chpl
iter abc() {
  for 1..8 do
    for j in 1..1 do
      yield j;
}
```
I then switched to a leader-follower one, but that didn't work quite well ([TIO4](https://tio.run/##fZExa8MwFIR3/YoHhUQCIWroEGI8ZexWsgfVUm3FimwkG6cx/u2untO4SaFd77473kl5KRttp8m02oN8zymDYSSkCxrepCv0ruxclRKy@I308gStLLaA2qtxikFfaq9RhCxbZGG1VDE1EAKQ1y604LrTXoYqQAYvKcoftYfWKDAOnoV4WvwherdQX/vKuOKAvZjM8SSaCLHhSyHHFpbOqU@jraIPKT5bIxn/H8LjPdbW/b404a9RVyJ2rFZ3tAjmopFL7uf@vvyHpwlb9j9Sqp5HoHHEZ4k7k5v4vQ2OKS6JiLQWqOFg1JkhfDENnT@RY27DYhB6H4utQ2y9hfUVTqfpCw)). I set the default value of `numTasks` to 4 so that the range gets divided evenly.
* The index `idx` starts from `2` instead of `1`
* A halt is produced: `halt reached - invoking orderToIndex on an integer 8 that is larger than the range's number of indices 8`

This leader-follower one should yield `1` eight times, but it fails and gives the error as mentioned in [TIO4](https://tio.run/##fZExa8MwFIR3/YoHhUQCIWroEGI8ZexWsgfVUm3FimwkG6cx/u2untO4SaFd77473kl5KRttp8m02oN8zymDYSSkCxrepCv0ruxclRKy@I308gStLLaA2qtxikFfaq9RhCxbZGG1VDE1EAKQ1y604LrTXoYqQAYvKcoftYfWKDAOnoV4WvwherdQX/vKuOKAvZjM8SSaCLHhSyHHFpbOqU@jraIPKT5bIxn/H8LjPdbW/b404a9RVyJ2rFZ3tAjmopFL7uf@vvyHpwlb9j9Sqp5HoHHEZ4k7k5v4vQ2OKS6JiLQWqOFg1JkhfDENnT@RY27DYhB6H4utQ2y9hfUVTqfpCw). I can't understand why this happened. Also, to check if the range was getting divided evenly and the number of yielded `1`s were eight, I wrote the following code (without an upper bound in the follower of the `forall` loop): [TIO5](https://tio.run/##fVJNT4NAFLzvr5jEpIWEUEk8mJKePHozvZst@yxbtgvZBdE2/e24j1psTfQ6H4@ZWYpSNmSGQbfkIDdFFON4EqLzhBdpt/RUdrbKhZj4Rjq5Ryu3SzD2rK2K0ZfkiEGsVhOcGpIquI5CAEVtfQvb7dfSVx4rPOQMv9UOrVbQFvdpejfxx8BdTH3tKm23r3yXnQVHirI0fUymgwlfifPR9anJqOjGlYzUSZz@L5KEPMbU/brU/q9SZ0W4MZtdqVOvD8S67Lru7@Q/@igbE/UuUMbeho1zYLFAW4emVFQhh2zh@DGgPTYUhHBUkH4nddnw9kuqHodgYsfThq2yC/i9D3Y5rxEk0hhEOoFWHzGLD7qJxh8hYV8cfFPMoJovMT9r82H4Ag).
* The number of yielded `1`s are 8, and so I can't understand what caused a problem in [TIO4](https://tio.run/##fZExa8MwFIR3/YoHhUQCIWroEGI8ZexWsgfVUm3FimwkG6cx/u2untO4SaFd77473kl5KRttp8m02oN8zymDYSSkCxrepCv0ruxclRKy@I308gStLLaA2qtxikFfaq9RhCxbZGG1VDE1EAKQ1y604LrTXoYqQAYvKcoftYfWKDAOnoV4WvwherdQX/vKuOKAvZjM8SSaCLHhSyHHFpbOqU@jraIPKT5bIxn/H8LjPdbW/b404a9RVyJ2rFZ3tAjmopFL7uf@vvyHpwlb9j9Sqp5HoHHEZ4k7k5v4vQ2OKS6JiLQWqOFg1JkhfDENnT@RY27DYhB6H4utQ2y9hfUVTqfpCw)
* The range being sent and the one being received is the same, so no problems there
* `idx` still starts from `2` for some reason


And I guess the initial error which I got a few days back (`zippered iterations have non-equal lengths`), [TIO6](https://tio.run/##lVFNa8JAEL3nVwwUTAJL0J5Kg0Jrb@1BiveyJGOyuNmVTWyq4m9PZ1ZNVWyhe5z3MfPeZqVcoe66dY3wLk2B03JtlmkQqAYdZJtMYyRdIaBRFdYwhmEMu/0FvJJOVtDI4hF4@qpMLuBa05TOtnUA9NoSHTIfxuNekWiUOVnuAuJ8SgdmXc1lvWS5WkAuGzmTzk9m6N5sJjXCBIZkjOYXGDWF4mVJJb8YI0rK/mTY20@OV3qfs6V@6tmZXVgntaZRDsrAMEnueuLOR8qsqRtorVsqU3xwJrbIuMqI6d5M9PbcTB6nXrpRqPPoQio8tA/@XbOAhdXatvNS1fEfXR9Y5DwYnCmSWm2ReaPDL9wO9cOPRrHvh9q5YuXWL2cANVZoGu6Nbj0hx9gnNOWox5Ij1AKoHVZs1So6hA@fnqcvoYD7WMCICn2IyQpaR/u08ZIQQq9Lu@4b), was related to this as well?

Am I doing something wrong here?

**Comment by bradcray** [comment]
Most of our standard leader-follower iterators yield a tuple of indices using 0-based indexing.  This is simply a convention so that if you're zippering something that's 0..7 with something that's 1..8 with something that's 1..16 by 2, they all have a common basis for what global iteration you're on.  This means that an 8-element range's follower is expecting to receive indices in the 0..7 range.  I expect that things are going wrong because your leader is yielding indices in the 1..8 range (but am surprised that the range iterator doesn't complain more about being given an iteration that's OOB... that seems like a bug).  Specifically, if I make your leader return working_iters-1, it seems to work as expected:  [TIO](https://tio.run/##fZExb4MwFIR3/4qnVEqM5FpF6hAFMXVspyp75WIXHByDMIg2iN9O/YDQJFK6vvvu8B1JJkplhkHXqgLxmdAAup6Qxil4FzZVL1lj84iQRS9FJY5Qi3QHeHvVVgbQZqpSeIQ4Xs7cKCG9qyMEICmsq8E2x71wuYMYniM8fxUV1FqCtvDE@cOid147m9qiyrVNPzAXnQk@iYacb9kSyDAliEZXW3nSWLp6m77/o5WRPgBW7DrrMZwdI0FvNDaKPen/L898B2OKdp9pd2@IifAZ6/UFzZ0@KeTCy4lu2/7xFJ87b3ZNyWKsgcIBp/TbhOfj3A4OETbxiDAGqGag5XeA8EmXdPzxDH3bwBuXCT222cFmgqNh@AU)

**Comment by akshansh2000** [comment]
@bradcray, yes, I should've taken care of that. I understand now.
Thank you for the explanation!

**Comment by bradcray** [comment]
It would've been much easier to take care of if the range follower had complained at you that you were requesting unreasonable indices...  It'd be good for one of us to look into why that was.

**Comment by bradcray** [comment]
> It'd be good for one of us to look into why that was.

I just did this and only now realized that it was complaining at you, athough the error message could've been clearer:

```
error: halt reached - invoking orderToIndex on an integer 8 that is larger than the range's number of indices 8
```

I missed it the first time because I forgot that TIO separates stdout from stderr.  Running it from my console made it much clearer...

**Comment by akshansh2000** [comment]
@bradcray, yes, it is warning me in the case of a `BoundedRange`, as in [TIO4](https://tio.run/##fZExa8MwFIR3/YoHhUQCIWroEGI8ZexWsgfVUm3FimwkG6cx/u2untO4SaFd77473kl5KRttp8m02oN8zymDYSSkCxrepCv0ruxclRKy@I308gStLLaA2qtxikFfaq9RhCxbZGG1VDE1EAKQ1y604LrTXoYqQAYvKcoftYfWKDAOnoV4WvwherdQX/vKuOKAvZjM8SSaCLHhSyHHFpbOqU@jraIPKT5bIxn/H8LjPdbW/b404a9RVyJ2rFZ3tAjmopFL7uf@vvyHpwlb9j9Sqp5HoHHEZ4k7k5v4vQ2OKS6JiLQWqOFg1JkhfDENnT@RY27DYhB6H4utQ2y9hfUVTqfpCw), however, I still do not get any error in the case of an `UnboundedRange`, as in [TIO5](https://tio.run/##fZExa8MwFIR3/YoHhUQCIWroUGI8dexWshfVUm3FimQkG7cx/u2untO4SaFd77473kllLVtt59l0OoB8KymDcSKkjxpepKv0U927Jidk9VsZ5BE6We0AtWfjFIOh1kGjCEWxysJqqVJqJASg9C524PrjXsYmQgEPOcrvPkBnFBgH90Lcrf6YvEto8KExrnrFXkyWeBLNhHjkayHHFpYvqU@jraI3Kb5YE5n@H8LTPdb6YV@b@NeoM5E6NpsrWkRz0shl13N/X/7D04yt@28p5ZcRaBzwWdLO7CJ@b4NDjksSIq0FajgY9cEQPpmWLp/IMcdSDoaQeq1DaruD7ZnN5/kL) (not even on my console).

**Comment by bradcray** [comment]
I think that's as expected.  Unbounded ranges are treated a bit specially in that they are permitted to be zippered with anything without running into a length check validation failure.  So since TIO5 wasn't 0-basing its indices, it yielded its second through #8'th element (so 2..9).

**Comment by akshansh2000** [comment]
I'd love to help with it in any way possible!

**Comment by bradcray** [comment]
@akshansh2000: I wish I knew how to advise you to do so.  It's a pretty big design challenge and likely to end up with a very different world than the one we have today.  The general ideas we've had (that I'm aware of) are based on increasing the amount of interaction between a leader iterator and the follower iterators such that it's up to a follower to say "mismatched zipper iteration" or just let the follower consume as much of the leader as it wants.  The thinking is that this interaction would also help fix a longstanding bug in which expressions like `forall (i,j) in zip(1..3, 1..4)` don't currently result in size mismatch errors.  (I took a stab at this latter bug in https://github.com/chapel-lang/chapel/issues/11685 and there may be some other helpful notes there as well... or not).  Again: big design challenge.

**Comment by akshansh2000** [comment]
@bradcray, I guess I'd have to dig deep into the core concepts of Chapel for that, on it!

**Comment by ben-albrecht** [comment]
For anyone curious, the current progress of the itertools library can be found here: https://github.com/chapel-lang/chapel/tree/master/test/library/packages/Itertools

Much work remains to be done.

**Comment by jabraham17** [comment]
I haven't read most of this very long issue, but found myself wanting an itertools package in Chapel and found this issue.

While writing Advent of Code solutions, I have written the following pattern several times

```chapel
iter allPairs(pts: []) do
  for i in 0..<pts.size do
    for j in i+1..<pts.size do
      yield (pts[i], pts[j]);
```

This is essentially what [itertools.combination](https://docs.python.org/3/library/itertools.html#itertools.combinations) might do in Python. I would love to have an Itertools package in Python that would avoid me having to write this boilerplate all the time.

This would also have some distinct advantages over my above code
* I keep writing `allPairs` as a serial iterator because I'm lazy, having a library that just takes care of the parallel iteration for me would be awesome. Not writing a parallel iterator myself means many of my solutions to AoC are parallel, not because they aren't parallelizable but because its too much code to parallelize for the level effort I want to put in.
* this is limited to pairs, a good library version would be agnostic of the "choose" part of the computation, i.e. this is `N choose 2`, but can we generalize to `N choose k`
* Possiblly some simple cases could avoid the need for `pts` to be an array, could it just be an arbitrary iterable? This would avoid some array creation overload and allow more "streaming"

**Comment by DanilaFe** [comment]
> this is limited to pairs, a good library version would be agnostic of the "choose" part of the computation, i.e. this is N choose 2, but can we generalize to N choose k

You can implement this in a parallel but suboptimal way as follows:

```Chapel
proc reordered(tup) {
  for param i in 0..(tup.size-2) do if tup[i] >= tup[i+1] then
    return true;
  return false;
}

proc repeat(x, param count) {
  if count < 1 then compilerError("don't do that");
  if count == 1 then return (x,);
  return (x, (...repeat(x, count-1)));
}

proc indexAll(A, x, param idx=0) {
  if x.size < 1 || idx < 0 || idx >= x.size then compilerError("don't do that");
  if idx == x.size-1 then return (A[x[idx]],);
  return (A[x[idx]], (...indexAll(A, x, idx+1)));
}


// A choose 2
forall c in {(...repeat(0..<A.size, 2))} do
  if !reordered(c) then writeln(indexAll(A, c));

// A choose 3
forall c in {(...repeat(0..<A.size, 3))} do
  if !reordered(c) then writeln(indexAll(A, c));

// A choose 4
forall c in {(...repeat(0..<A.size, 4))} do
  if !reordered(c) then writeln(indexAll(A, c));
```

This works by iterating the (A.size)^n space of indices, in parallel over the outermost index, rejecting re-ordered indices (`1,2,3` ok, `1,3,2` not ok), and then indexing into `A` using these. There are two ways in which this is suboptimal.

1. Instead of configuring iteration such that we don't visit the re-oredered indices, we iterate over them and rule them out as part of the computation. This wastes cycles, but has the convenient effect of not having to write diagonal iteration algorithms. 
2. The parallelization does not distribute the workload evenly. Consider the simple case of `pairs` with an array of size `4*n`, parallelized 4 ways. For the elements assigned to "chunk 1", they have to yield pairs from all the remaining chunks, for something like `3n^2` (across chunk boundary) + `n(n-1)/2` (within chunk) yields. On the other hand, the elements in the last chunk only need to yield `n(n-1)^2/2` (within chunk) times, and 0 for "across chunk". In other words, the threads for earlier chunks have to do linearly more work. I think this becomes quadratic for 3-element combinations, etc.

To work around problem (2), I think we need to parallelize by taking equally-sized diagonal slices of the A.size^n index space, but how to do that with fixed overhead is not immediately obvious to me, though I can maybe envision some ideas.

**Comment by DanilaFe** [comment]
Though suboptimal, a parallel `pairs` algorithm does produce a speedup compared to the serial case.

```Chapel
iter pairs(input): 2*input.eltType {
  for i in 0..<input.size {
    for j in i+1..<input.size {
      yield (input[i], input[j]);
    }
  }
}

iter parallelPairs(input: []): 2*input.eltType {
  halt("no"); 
}

iter parallelPairs(input: [], param tag) where tag == iterKind.standalone {
  forall c in {(...repeat(0..<input.size, 2))} do
    if !reordered(c) then yield indexAll(input, c);
}

config const problemSize = 100;
var A = foreach i in 0..<problemSize do i;
var ser = + reduce pairs(A);
var par = + reduce parallelPairs(A);
```

With some logging and a problem size of 20000, the output was:

```
(1333133340000, 2666466670000) (elapsed: 5.68142 seconds)
(1333133340000, 2666466670000) (elapsed: 1.81556 seconds)
Speedup: 3.1293
```

This was with `here.numTasksPar == 8`, so, far from ideal, but measurable.

**EDIT**

I wrote a properly parallel version:
```Chapel
iter parallelPairs(input: [], param tag) where tag == iterKind.standalone {
  use RangeChunk, Math;
  const idxSpace = 0..<((input.size * (input.size - 1)) / 2);
  coforall chunk in chunks(idxSpace, here.maxTaskPar) {
    var startIdx = chunk.low,
        endIdx = chunk.high;

    var col = ((floor(sqrt(8*startIdx+1)) : int)-1) / 2;
    var colStart = (col * (col + 1)) / 2;
    var row = startIdx - colStart;

    while startIdx <= endIdx {
      yield (input[row], input[col+1]);
      startIdx += 1;
      row += 1;
      if row > col {
        col += 1;
        row = 0;
      }
    }
  }
}
```
Across the three version:

```
(10665866680000, 21332533340000) (elapsed: 23.3616 seconds)
(10665866680000, 21332533340000) (elapsed: 7.48834 seconds)
(10665866680000, 21332533340000) (elapsed: 3.79388 seconds)
Speedup: 3.11972 6.15769
```

6.15 is still not the ideal speedup of 8, but it's much closer.

**EDIT 2**

Embarrassingly, I measured the speedups without `--fast`. With `--fast`, neither version performs faster than the serial. 🤦
